Cookies and arguments parsing improvements

stan1y opened the pull request:
## Cookies support
Functions added:
* http_populate_cookies - parse cookies from a header into request's TAILQ
* http_request_cookie - get a cookie by name from request's TAILQ
* http_response_cookie - creates a new cookie and initialises it with defaults. before returning to caller, adds a new cookie to response's TAILQ.

## Cookies format
Cookies can be sent with all features specified in RFC 6265. The formatting of generated cookies is based on  [MDN Set-Cookie header spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Set-Cookie) and parsing on [MDN Cookie header spec](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie)

## Usage
`http_response_cookie` can be used together with `http_response_header` as illustrated in the example. Each call to `http_response_cookie` would produce a `set-cookie` header in the response.

jorisvink commented:
> Change: functions now return size_t of parsed arguments, instead of void.

Heh that is kind of funny as I actually moved away from returning this as I found it gave no useful information to the caller. So what if you know there's 5 parameters that parsed correctly? You still would need to look at each and every one of them to know if they are present or not.

But if we're bringing this back size_t is the wrong type for these return values and they should probably just be an int.

> 4f9044c
> Is not required. Sizes are in fact "size_t".

Except that is not what PQexecParams() takes. It takes an int for lengths and thus we make sure we provide this higher up as well to avoid confusion.

The rest of the new additions look sensible and I'll take a deeper look under the day! :)

stan1y commented:
>So what if you know there's 5 parameters that parsed correctly? You still would need to look at each and every one of them to know if they are present or not.
I application expects 6 parameters and only 5 were parsed there is no need to iterate over list to tell which exactly were not sent and just send code 400.
The rationale is to let application make a decision if iteration of parsed arguments worth it.

jorisvink commented on src/http.c:
```diff
@@ -925,17 +983,71 @@ http_file_rewind(struct http_file *file)
+
+    header = kore_strdup(c);
+    if (strstr(header, semicol) != NULL) {
+    	delim = semicol;
```
Does the RFC allow for anything else besides ';' as a cookie delim?

stan1y replied:
There are several RFCs on subject and they do not restrict it explicitly. 
RFC 2109 was obsoleted by RFC 2965, which was obsoleted by RFC 6265.
A lot of browsers actually did (or still do), see https://bugzilla.mozilla.org/show_bug.cgi?id=370639.
The convention is ";" however "," is explicitly mentioned in RFC 2109. Most servers and clients support  RFC 2109 AFAIK.

jorisvink replied:
Browsers may still support parsing incoming cookies like this but I'm hopeful they are not going to send comma delimited Set-cookie headers asRFC 6265 does not define this at all.

Either way I'd kill off the comma and all the magic to figure out what the delimiter here is.

stan1y replied:
Ok, agree.  Also, do we want to check name/value content for anything prohibited by RFCs? Also see http://stackoverflow.com/questions/1969232/allowed-characters-in-cookies

jorisvink replied:
We currently don't do that in the headers either, so skip that for now. It is on my todo list for this to happen.

jorisvink commented on src/http.c:
```diff
@@ -925,17 +983,71 @@ http_file_rewind(struct http_file *file)
+		ck = kore_pool_get(&http_cookie_pool);
+		ck->name = kore_malloc(sizeof(char) * (nlen + 1));
+		memset(ck->name, 0, sizeof(char) * (nlen + 1));
+		strncpy(ck->name, c, nlen);
```
No.

Please don't ever do this.

You can easily replace the '=' character with a NUL-byte and just use kore_strdup() on it for the name instead of this error-prone idiom.

jorisvink commented on src/pgsql.c:
```diff
@@ -169,7 +169,7 @@ kore_pgsql_v_query_params(struct kore_pgsql *pgsql,
 		for (i = 0; i < count; i++) {
 			values[i] = va_arg(args, void *);
-			lengths[i] = va_arg(args, int);
+			lengths[i] = va_arg(args, size_t);
```
No.

This breaks kore_pgsql_query_params().

jorisvink commented:
Any new progress on this?

The cookie support is definitely solid and I'd like to get this in.

stan1y commented:
I am busy with my actual job. Your concerns would be resolved on weekend

jorisvink commented:
Neat.

stan1y commented:
Anything else required for merge?

jorisvink commented:
An example app for cookie usage would be good to have!

jorisvink commented on src/http.c:
```diff
@@ -957,6 +1015,43 @@ http_file_rewind(struct http_file *file)
+			kore_log(LOG_ERR, "cookie name is too long");
+			continue;
+		}
+		value = "\0";
```
This seems very wrong.

You are pointing value to an empty C string which is then used in kore_strdup() for the name member.

After that you increase the pointer by one position (value++) and store whatever value is pointing towards now in the value member using kore_strdup() again.

I guess you ment something more like:

```
*(value)++ = '\0';
ck->name = kore_strdup(c);
ck->value = kore_strdup(value);
```

jorisvink commented on src/http.c:
```diff
@@ -957,6 +1015,43 @@ http_file_rewind(struct http_file *file)
+			kore_log(LOG_ERR, "malformed cookie");
+			continue;
+		}
+		nlen = value - c - 1;
```
Careful here, what happens to nlen when value == c?

stan1y commented:
I've removed manual parsing in favour of `kore_split_string` and added RFC 6265 feature to the response cookies. Example added as well.
In a separate branch I am building stand-alone testsuite for kore with [py.test](http://doc.pytest.org/en/latest/)

jorisvink commented:
The build is currently failing so I'm holding off to look at it until all green.

stan1y commented:
all green

jorisvink commented:
Neat

jorisvink commented on examples/cookies/README.md:
```diff
@@ -0,0 +1,11 @@
+Generic Kore application that shows off a few things:
```
The example README could use some changes as this seems copy/pasted from examples/generic.

jorisvink commented on examples/cookies/src/example.c:
```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2013 Joris Vink <joris@coders.se>
```
I didn't write this, you're the author of this code.

jorisvink commented on includes/http.h:
```diff
@@ -63,6 +67,22 @@ struct http_header {
+	char			*value;
+	char			*path;
+	char			*domain;
+	unsigned int	 maxage;
```
Please use exact width integer types.

jorisvink commented on src/http.c:
```diff
@@ -471,6 +484,28 @@ http_request_free(struct http_request *req)
+		TAILQ_REMOVE(&(req->resp_cookies), ck, list);
+		kore_free(ck->name);
+		kore_free(ck->value);
+		if (ck->path != NULL)
```
kore_free() knows how to deal with NULL pointers so no need to check for NULL before calling it.

jorisvink commented on src/http.c:
```diff
@@ -593,6 +628,21 @@ http_request_header(struct http_request *req, const char *header, char **out)
 }
 
 int
+http_request_cookie(struct http_request *req, const char* cookie, char **out)
```
small style nitpick on the * here.

jorisvink commented on src/http.c:
```diff
@@ -956,6 +1006,61 @@ http_file_rewind(struct http_file *file)
+	if (name == NULL || strlen(name) == 0 ||
+		val == NULL || strlen(val) == 0) {
+		kore_log(LOG_ERR, "invalid cookie values to set");
+		return NULL;
```
style nitpicks, add parenthesis around return statements like everywhere else.

jorisvink commented on src/http.c:
```diff
@@ -956,6 +1006,61 @@ http_file_rewind(struct http_file *file)
+	ck->maxage = UINT32_MAX;
+
+	TAILQ_INSERT_TAIL(&(req->resp_cookies), ck, list);
+	return ck;
```
same here.

jorisvink commented on src/http.c:
```diff
@@ -956,6 +1006,61 @@ http_file_rewind(struct http_file *file)
+		for (c = cookies[i]; isspace(*c); c++)
+			;
+
+		n = kore_split_string(c, "=", pair, 3);
```
Off by one here, pair holds 2 pointers not 3.

Since kore_split_string() requires a guard parameter you'll just want to bump pair to hold 3.

jorisvink commented on src/http.c:
```diff
@@ -956,6 +1006,61 @@ http_file_rewind(struct http_file *file)
+
+		n = kore_split_string(c, "=", pair, 3);
+		if (n != 2) {
+			kore_log(LOG_ERR, "malformed cookie: %s", c);
```
I'd rather not have this log anything in case people are spamming requests with malformed cookies.

jorisvink commented on src/http.c:
```diff
@@ -1488,6 +1597,32 @@ http_response_normal(struct http_request *req, struct connection *c,
+			if (ck->domain != NULL)
+				kore_buf_appendf(ckbuf, "; Domain=%s", ck->domain);
+			if (ck->expires > 0) {
+				strftime(expires, HTTP_DATE_MAXSIZE, "%a, %d %b %y %H:%M:%S GMT",
```
strftime() can fail, probably wants a return value check.

we probably want the reentrant version of gmtime() so that we never run into trouble if TASKS are enabled.

jorisvink replied:
Additionally I'd like to add that using HTTP_DATE_MAXSIZE here is not obvious it relates to the size of expires.

Since it is in scope sizeof(expires) as the length is much more readable and removes any doubt that it is the length of the expires array.

jorisvink commented on src/http.c:
```diff
@@ -1488,6 +1597,32 @@ http_response_normal(struct http_request *req, struct connection *c,
+				kore_buf_appendf(ckbuf, "; Expires=%s", expires);
+			}
+			if (ck->maxage != UINT32_MAX) {
+				kore_buf_appendf(ckbuf, "; Max-Age=%d", ck->maxage);
```
maxage is unsigned but %d specifies a signed integer.

jorisvink commented on src/http.c:
```diff
@@ -54,12 +54,14 @@ static int	multipart_parse_headers(struct http_request *,
 		    const char *, const int);
 
 static struct kore_buf			*header_buf;
+static struct kore_buf			*ckhdr_buf;
```
ckhdr_buf appears unused now?

jorisvink commented on src/http.c:
```diff
@@ -956,6 +1006,61 @@ http_file_rewind(struct http_file *file)
+		ck = kore_pool_get(&http_cookie_pool);
+		ck->name = kore_strdup(pair[0]);
+		ck->value = kore_strdup(pair[1]);
+		TAILQ_INSERT_TAIL(&(req->req_cookies), ck, list);
```
Is there a need to parse incoming cookie params such as Secure and HttpOnly?

jorisvink commented:
OK - i went through it and left some comments for small improvements.

The overall approach and implementation I'm happy with, if you could just fix those small nitpicks that would be fantastic.

stan1y commented:
Style and other fixes uploaded.

>Is there a need to parse incoming cookie params such as Secure and HttpOnly?

No. Such flags are available only in "Set-Cookie" header.

jorisvink commented:
Excellent.
